nemo/collections/asr/data/audio_to_text.py

```diff
@@ -182,3 +160,160 @@ def _collate_fn(self, batch):
+        return {
+            'audio_signal': NeuralType(
+                ('B', 'T'),
+                AudioSignal(freq=self._sample_rate)
+                if self is not None and hasattr(self, '_sample_rate')
+                else AudioSignal(),
+            ),
+            'a_sig_length': NeuralType(tuple('B'), LengthsType()),
+            'transcripts': NeuralType(('B', 'T'), LabelsType()),
+            'transcript_length': NeuralType(tuple('B'), LengthsType()),
+        }
+
+    def __init__(
+        self,
+        manifest_filepath,
+        labels,
+        featurizer,
+        max_duration=None,
+        min_duration=None,
+        max_utts=0,
+        blank_index=-1,
+        unk_index=-1,
+        normalize=True,
+        trim=False,
+        bos_id=None,
+        eos_id=None,
+        pad_id=0,
+        load_audio=True,
+        parser='en',
+        add_misc=False,
+    ):
+        self.labels = labels
+
+        parser = parsers.make_parser(
+            labels=labels, name=parser, unk_id=unk_index, blank_id=blank_index, do_normalize=normalize,
+        )
+
+        super().__init__(
+            manifest_filepath=manifest_filepath,
+            featurizer=featurizer,
+            parser=parser,
+            max_duration=max_duration,
+            min_duration=min_duration,
+            max_utts=max_utts,
+            trim=trim,
+            bos_id=bos_id,
+            eos_id=eos_id,
+            pad_id=pad_id,
+            load_audio=load_audio,
+            add_misc=add_misc,
+        )
+
+
+@experimental
+class AudioToBPEDataset(_AudioDataset):
+    @property
+    def output_types(self) -> Optional[Dict[str, NeuralType]]:
+        """Returns definitions of module output ports.
+               """
+        return {
+            'audio_signal': NeuralType(
+                ('B', 'T'),
+                AudioSignal(freq=self._sample_rate)
+                if self is not None and hasattr(self, '_sample_rate')
+                else AudioSignal(),
+            ),
+            'a_sig_length': NeuralType(tuple('B'), LengthsType()),
+            'transcripts': NeuralType(('B', 'T'), LabelsType()),
+            'transcript_length': NeuralType(tuple('B'), LengthsType()),
+        }
+
+    def __init__(
+        self,
+        manifest_filepath,
+        tokenizer,
+        featurizer,
+        max_duration=None,
+        min_duration=None,
+        max_utts=0,
+        trim=False,
+        load_audio=True,
+        add_misc=False,
+    ):
+        if hasattr(tokenizer, 'bos_token'):
+            bos_id = tokenizer.bos_id
+        else:
+            bos_id = None
+
+        if hasattr(tokenizer, 'eos_token'):
+            eos_id = tokenizer.eos_id
+        else:
+            eos_id = None
+
+        if hasattr(tokenizer, 'pad_token'):
+            pad_id = tokenizer.pad_id
+        else:
+            pad_id = 0
+
+        class TokenizerWrapper:
+            def __init__(self, tokenizer):
+                self._tokenizer = tokenizer
```

Comment:
Let's add a check to make sure that `tokenizer` is a child of TokenizerSpec

Reply:
Good point.

Reply:
I actually cant check this without first moving this to another file (cause import would import from NLP side)

Reply:
You are already doing thie line `from nemo.collections.common.tokenizers.tokenizer_spec import TokenizerSpec` inside of `nemo/collections/asr/metrics/wer_bpe.py`, can you explain what you mean by "cause import would import from NLP side"?
From what I see, ASR is already importing TokenizerSpec, so why not just use the import liberally at this point?

Reply:
This file does not import anything from NLP, nor does the __init__ export EncDecAsrModelBPE (so just importing the Nemo collection will not import anything from NLP side). You would need to explicit import the BPE side and built a model with that, and that's why the user script imports from two different files.

Reply:
This file imports `nemo/collections/asr/metrics/wer_bpe.py` which in turn imports `from nemo.collections.common.tokenizers.tokenizer_spec import TokenizerSpec`, so you are already importing `TokenizerSpec` to load this file, why not add the check?

Reply:
Nvm, I was looking at the wrong file.